ExtractUserDataActivity/handler.ts

```diff
@@ -0,0 +1,445 @@
+              kind: "USER_NOT_FOUND_FAILURE"
+            })
+          )(maybeProfile)
+        )
+    );
+
+  /**
+   * Retrieves all contents for provided messages
+   * @param messages
+   */
+  const getAllMessageContents = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<MessageContentWithId>
+  > =>
+    array.sequence(taskEither)(
+      messages.map(({ id: messageId }) =>
+        fromQueryEither(
+          () => messageModel.getContentFromBlob(blobService, messageId),
+          "messageModel.getContentFromBlob"
+        ).foldTaskEither<ActivityResultQueryFailure, MessageContentWithId>(
+          failure => fromEither(left(failure)),
+          maybeContent =>
+            fromEither(
+              fromOption(
+                ActivityResultQueryFailure.encode({
+                  kind: "QUERY_FAILURE",
+                  query: "messageModel.getContentFromBlob",
+                  reason: `Cannot find content for message ${messageId}`
+                })
+              )(maybeContent).map<MessageContentWithId>(
+                (content: MessageContent) => ({
+                  content,
+                  messageId
+                })
+              )
+            )
+        )
+      )
+    );
+
+  /**
+   * Given a list of messages, it queires for relative notifications
+   * @param messages
+   */
+  const findNotificationsForAllMessages = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<RetrievedNotification>
+  > =>
+    array
+      .sequence(taskEither)(
+        messages.map(m =>
+          fromQueryEither<ReadonlyArray<RetrievedNotification>>(
+            () =>
+              iteratorToArray(
+                notificationModel.findNotificationsForMessage(m.id)
+              ),
+            "findNotificationsForRecipient"
+          )
+        )
+      )
+      .foldTaskEither(
+        e => fromEither(left(e)),
+        arrayOfArray =>
+          // tslint:disable-next-line: readonly-array
+          fromEither(right(flatten(arrayOfArray as RetrievedNotification[][])))
+      );
+
+  const findAllNotificationStatuses = (
+    notifications: ReadonlyArray<RetrievedNotification>
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<NotificationStatus>
+  > =>
+    array
+      .sequence(taskEither)(
+        // compose a query for every supported channel type
+        notifications
+          .reduce(
+            (queries, { id: notificationId }) => [
+              ...queries,
+              ...Object.values(NotificationChannelEnum).map(channel => {
+                switch (channel) {
+                  case NotificationChannelEnum.EMAIL:
+                  case NotificationChannelEnum.WEBHOOK:
+                    return [notificationId, channel];
+                  default:
+                    assertNever(channel);
```

Comment:
cannot `...Object.values(NotificationChannelEnum).map(channel => [notificationId, channel])` 
just work without the switch here?

Reply:
Yes, but this way I could implement an exahustive check that will force us to update this function if in the future we add/remove options in `NotificationChannelEnum `

Reply:
why do we want to be forced to update this method in case?

Reply:
Maybe we'll add support for another notification channel

Reply:
and why `.map(channel => [notificationId, channel])` won't suffice in that case?